crates/turborepo-scm/src/ls_tree.rs

```diff
@@ -0,0 +1,143 @@
+use std::{
+    io::{BufRead, BufReader, Read},
+    process::{Command, Stdio},
+};
+
+use nom::Finish;
+use turbopath::{AbsoluteSystemPathBuf, RelativeUnixPathBuf};
+
+use crate::{package_deps::GitHashes, read_git_error, wait_for_success, Error};
+
+pub fn git_ls_tree(root_path: &AbsoluteSystemPathBuf) -> Result<GitHashes, Error> {
+    let mut hashes = GitHashes::new();
+    let mut git = Command::new("git")
+        .args(["ls-tree", "-r", "-z", "HEAD"])
+        .current_dir(root_path)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    let stdout = git
+        .stdout
+        .as_mut()
+        .ok_or_else(|| Error::git_error("failed to get stdout for git ls-tree"))?;
+    let mut stderr = git
+        .stderr
+        .take()
+        .ok_or_else(|| Error::git_error("failed to get stderr for git ls-tree"))?;
+    let result = read_ls_tree(stdout, &mut hashes);
+    if let Err(err) = result {
+        return Err(read_git_error(&mut stderr).unwrap_or(err));
+    }
+    wait_for_success(git, &mut stderr, "git ls-tree", root_path)?;
+    Ok(hashes)
+}
+
+fn read_ls_tree<R: Read>(reader: R, hashes: &mut GitHashes) -> Result<(), Error> {
+    let mut reader = BufReader::new(reader);
+    let mut buffer = Vec::new();
+    loop {
+        buffer.clear();
+        {
+            let bytes_read = reader.read_until(b'\0', &mut buffer)?;
+            if bytes_read == 0 {
+                break;
+            }
+            {
```

Comment:
What's the reason for this being in a block?

Reply:
I think an artifact from an earlier iteration. Removed.